qiskit_ode/signals/signals.py

```diff
@@ -231,24 +277,75 @@ def __init__(
             dt: The duration of each sample.
             samples: The array of samples.
             start_time: The time at which the signal starts.
-            duration: The duration of the signal in samples.
             carrier_freq: The frequency of the carrier.
             phase: The phase of the carrier.
             name: name of the signal.
         """
-        super().__init__(name)
-
+        self._name = name
+        self._is_constant = False
         self._dt = dt
+        self._samples = Array(samples)
+        self._start_time = start_time
 
-        if samples is not None:
-            self._samples = Array(samples)
-        else:
-            self._samples = Array([0.0] * duration)
+        # initialize internally stored carrier/phase information
+        self._carrier_freq = None
+        self._phase = None
+        self._carrier_arg = None
+        self._phase_arg = None
 
-        self._start_time = start_time
+        # set carrier and phase
+        self.carrier_freq = carrier_freq
+        self.phase = phase
 
-        self._carrier_freq = Array(carrier_freq)
-        self._phase = Array(phase)
+    @classmethod
+    def from_Signal(
+        cls,
+        signal: Signal,
+        dt: float,
+        n_samples: int,
+        start_time: Optional[float] = 0.0,
+        sample_carrier: Optional[bool] = False,
+    ):
+        r"""Constructs a ``DiscreteSignal`` object by sampling a ``Signal``\.
+
+        The optional argument ``sample_carrier`` controls whether or not to include the carrier
+        in the sampling. I.e.:
+
+            - If ``sample_carrier == False``\, a ``DiscreteSignal`` is constructed with:
+                - ``samples`` obtained by sampling ``signal.envelope``\.
+                - ``carrier_freq = signal.carrier_freq``\.
+                - ``phase = signal.phase``\.
+
+            - If ``sample_carrier == True``\, a ``DiscreteSignal`` is constructed with:
+                - ``samples`` obtained by sampling ``signal`` (as a ``callable``\)
+                - ``carrier_freq = 0``\.
+                - ``phase = signal.phase``\.
+
+        In either case, samples are obtained from the midpoint of each interval.
+
+        Args:
+            signal: Signal to sample.
+            dt: Time increment to use.
+            n_samples: Number of steps to resample with.
+            start_time: Start time from which to resample.
+            sample_carrier: Whether or not to include the carrier in the sampling.
+
+        Returns:
+            DiscreteSignal: The discretized ``Signal``\.
+        """
+
+        times = start_time + (np.arange(n_samples) + 0.5) * dt
+        freq = signal.carrier_freq
+
+        if sample_carrier:
+            freq = 0.0
+            samples = signal(times)
+        else:
+            samples = signal.envelope(times)
+
+        return DiscreteSignal(
+            dt, samples, start_time=start_time, carrier_freq=freq, phase=signal.phase
```

Comment:
Why no name?

Reply:
Not sure what name to assign to this:

- If the original signal has no name then I think it's fine to not assign a name.
- If the original signal does have a name, if we want to assign a name we could do something like `'Discretized(dt=0.1, name)'`? Or do you think we should just set the name to the name of `signal`?

Reply:
I think we can just set the name to the name of the signal.